Automattic/wp-calypso
# [nudge-a-palooza] Move redirection logic from controller to higher-order components

This PR revisits access control/redirection logic for upsell pages - it adds a few more redirects, and also moves them from controller level where state may not be fully loaded yet, to a higher-order component that is aware of every state change and may react accordingly. The reason is that the following bug was found:
1. User navigated to `/plugins` on a free site
1. They clicked on "upload plugin" and were redirected to an upsell page
1. They switched to a business site and still ended up on an upsell page

## Test plan:
**On calypso.localhost where upsell/nudge-a-palooza feature is enabled**
1. Go to a free site
1. Add yourself to `customPluginAndThemeLandingPages` group of nudgeAPalooza A/B test
1. Visit the following 4 pages and confirm each is loaded correctly:
* `/feature/plugins`
* `/feature/themes`
* `/feature/ads`
* `/feature/store`
1. On `/feature/plugins` switch to a premium site and confirm you end up on an upsell page
1. On `/feature/plugins` switch to a non-AT business site and confirm you end up on `/plugins`
1. On `/feature/plugins` switch to an AT business site and confirm you end up on `/plugins` or `/wp-admin/plugin-install.php`
1. On `/feature/themes` switch to a premium site and confirm you end up on an upsell page
1. On `/feature/themes` switch to a non-AT business site and confirm you end up on `/themes`
1. On `/feature/themes` switch to an AT business site and confirm you end up on `/themes`
1. On `/feature/ads` switch to a personal site and confirm you end up on an upsell page
1. On `/feature/ads` switch to a premium site and confirm you end up on wordads config or stats
1. On `/feature/ads` switch to a business site and confirm you end up on wordads config or stats
1. On `/feature/store` switch to a personal site and confirm you end up on an upsell page
1. On `/feature/store` switch to a premium site and confirm you end up on an upsell page
1. On `/feature/store` switch to a non-AT business site and confirm you end up on `/store`
1. On `/feature/store` switch to an AT business site and confirm you end up on `/store`
1. Go to a personal site that you are not eligible to upgrade and confirm that each of the 4 pages listed above redirects you to `/stats`
1. Go to `/plugins` on a free site and confirm that clicking on "Upload plugin" redirects to an upsell page
1. Go to `/plugins` on a business site and confirm that clicking on "Upload plugin" redirects to an upload page
1. Go to `/themes` on a free site and confirm that clicking on "Upload theme" redirects to an upsell page
1. Go to `/themes` on a business site and confirm that clicking on "Upload theme" redirects to an upload page
1. Add yourself to `control` group of `nudgeAPalooza` A/B test
1. Confirm you can click "Upload theme" and "Upload plugin" buttons on both free and business sites and no upsell page is displayed

**On calypso.live link below where nudge-a-palooza is disabled**
1. Add yourself to `customPluginAndThemeLandingPages` group of nudgeAPalooza A/B test
1. Confirm you can click "Upload theme" and "Upload plugin" buttons on both free and business sites and no upsell page is displayed
